cloudsync_onedrive.py

```diff
@@ -607,6 +483,8 @@ def _raise_converted_error(self, *, ex=None, req=None):      # pylint: disable=t
             raise CloudFileExistsError(msg)
         if code == ErrorCode.AccessDenied:
             raise CloudFileExistsError(msg)
+        if code == ErrorCode.NotSupported:
```

Comment:
is this new behavior?

Reply:
this was in the `upload` function before, I just moved it here.

`                except onedrivesdk.error.OneDriveError as e:
                    if e.code == ErrorCode.NotSupported:
                        raise CloudFileExistsError("Cannot upload to folder")
`

Reply:
Are you sure that `ErrorCode.NotSupported` can't happen other places besides `upload`? If we're moving it here, then it seems to be a much more generalized response to that error code, and if it can happen somewhere else, then perhaps `CloudFileExistsError` would not be the appropriate conversion

Reply:
good question, I doubt even Microsoft can answer it with any certainty.

my thinking was that if the same error code does come up in other contexts, currently it is converted to the generic unknown error of type Exception, which also isn't great.

not wholly opposed to moving this back to `upload()` though it would be our only such special case.

Reply:
If you want, you can go back to the old version of the code, remove the translation for `NotSupported` from the upload method, and run the provider tests to see which one fails. It's probably something like uploading a file over an existing folder, or something like that. Once you've pinned down what causes that, you can try to see if you can think of an equivalent invalid operation for download, like downloading a folder instead of a file for example, to see if you get `NotSupported` for that in download as well. If you are able to reproduce `NotSupported` in download, that probably means that there is a gap in our tests, and that needs to be added. At the very least, you could recover the lost reason why that error can come up, and document it in the code

Reply:
it is coming from uploading a file over an existing folder. download throws a different error (FNF).

interestingly, we do a pre-check for the upload-to-folder condition in the upload-with-data codepath, but not in the upload-zero-byte codepath.

ended up removing this from error conversion and refactoring `upload()` a bit to raise the error in both codepaths. Note: we could save an API hit by not doing the upload pre-checks and handling OneDrive errors instead, but this seems a little safer than relying on the OD api to not change.

Reply:
You raise an interesting point about safety. I think you're right about eliminating the upload pre-checks and handling OneDrive errors instead, actually, but for a better reason than eliminating the API hit, although that is a pretty good reason by itself. While the OD api may or may not be reliable, the problem with pre-checks are that they are never reliable, because it is always possible that the condition we pre-check for could appear after the pre-check and before the operation that depends on that check. At least, we should ensure that the translation is there, because the condition can still appear when the pre-check fails in this way. I'd be in favor of removing the pre-checks entirely and just going with the error translation.

The OD api can always change, so I don't think being defensive about that actually helps. Your solution depends on one portion of the api (in the pre-check) that "allows" us to avoid depending on the other portion of the api (what exception happens when), but if the api is suspect, then I'm not convinced that depending on one more than the other resolves the underlying concern that the behavior of the api might change. Given that, I'd recommend, ditching the pre-checks and restoring the exception translation.

Reply:
the prechecks use core pieces of the api (check if an oid exists, check if the oid represents a folder) - if those change, a lot of things will break. a change to the text of an error message (say "notSupported" to "unsupported") is much more subtle, so not all api changes are created equal.

happy to put the translations back, but leaning towards keeping the prechecks. thoughts?

Reply:
I agree that not all api changes are created equal, so I get why a concern about api stability would lean towards a pre-check. It's just that I'm not a fan of making an extra check that carries a cost, while implying a level of security that it doesn't actually provide.

Also, I think the fact that the error message is in text (versus some other form) would not imply a higher likelihood it will change. Since the response that comes back from the API is in json, everything in that response must be text. Any changes to the text in the json that departs from the documentation must be a breaking change. They'd be more likely to change that condition to a different error type than to change the way a particular error type is reported, IMO, and if they did change it to a different error type, we'd land in another error translation that would likely still result in similar behavior of the sync engine.

Reply:
Is there a way to remove the pre-check and handle the errors better? These pre-checks are fundamentally flawed, and as a result, the errors that you're seeing can still happen, but only more rarely. Pre-checks just pave over the problem and get the tests to pass, but that doesn't mean it's fixed

Reply:
i haven't found a way. onedrive (personal) throws the same exact FNF error (same message) in two different scenarios, but provider tests expect FEX in one of the two.

guessing this is the reason the pre-checks were there in the first place.

Reply:
I love the new post-checks. They avoid the expense of the double check, except during exceptional circumstances, and it prevents the race condition inherent in a pre-check. There is a _slight_ chance of a race condition if the files change between the error and the post-check. Is it possible to simply check if we are connected to personal or business, and interpret the error differently based on that? That seems the most deterministic

Reply:
per offline discussion, this is the only way to differentiate oid-not-found and oid-is-a-folder errors